group-coordinator/src/main/java/org/apache/kafka/coordinator/group/metrics/CoordinatorRuntimeMetrics.java

```diff
@@ -0,0 +1,244 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.coordinator.group.metrics;
+
+import com.yammer.metrics.core.Gauge;
+import com.yammer.metrics.core.MetricName;
+import com.yammer.metrics.core.MetricsRegistry;
+import org.apache.kafka.common.metrics.Metrics;
+import org.apache.kafka.common.metrics.Sensor;
+import org.apache.kafka.common.metrics.stats.Avg;
+import org.apache.kafka.common.metrics.stats.Max;
+import org.apache.kafka.server.metrics.KafkaYammerMetrics;
+
+import java.util.Arrays;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Used by the group and transaction coordinator runtimes, the metrics suite holds partition state gauges and sensors.
+ */
+public abstract class CoordinatorRuntimeMetrics implements AutoCloseable {
+    /**
+     * Metric to count the number of partitions in Initial state.
+     */
+    protected final MetricName numPartitionsInitial = getMetricName("NumPartitionsInitial");
+    /**
+     * Visible for testing.
+     */
+    public static final AtomicLong NUM_PARTITIONS_INITIAL = new AtomicLong(0);
+
+    /**
+     * Metric to count the number of partitions in Loading state.
+     */
+    protected final MetricName numPartitionsLoading = getMetricName("NumPartitionsLoading");
+    /**
+     * Visible for testing.
+     */
+    public static final AtomicLong NUM_PARTITIONS_LOADING = new AtomicLong(0);
+
+    /**
+     * Metric to count the number of partitions in Active state.
+     */
+    protected final MetricName numPartitionsActive = getMetricName("NumPartitionsActive");
+    /**
+     * Visible for testing.
+     */
+    public static final AtomicLong NUM_PARTITIONS_ACTIVE = new AtomicLong(0);
+
+    /**
+     * Metric to count the number of partitions in Closed state.
+     */
+    protected final MetricName numPartitionsClosed = getMetricName("NumPartitionsClosed");
```

Comment:
I wonder if there is any value is exposing Closed and Initial states. The partition never remains in Closed state. We transition to it, unload and remove the shard context. Or we should perhaps say Unloading in the metric... What do you think?

Reply:
removing initial makes sense to me. i'll also remove closed, since the counter will never decrease. creating a sensor for this also doesn't make much sense.